checks/cs50/2017/x/scratch/checks.py

```diff
@@ -0,0 +1,155 @@
+            if not can_meow:
+                return
+        
+        # If we haven't returned, then no non-cat sprite found.
+        raise Error("Requires a non-cat sprite.")
+
+    @check("valid")
+    def three_scripts(self):
+        """project contains at least three scripts"""
+        project = json.loads(File("project.json").read())
+        
+        num_scripts = 0
+
+        # Add up scripts from each sprite or backdrop.
+        for child in project["children"]:
+            if "scripts" in child:
+                num_scripts += len(child["scripts"])
+
+        if num_scripts < 3:
+            raise Error("Only {} script{} found, 3 required.".format(num_scripts,
+                "" if num_scripts == 1 else "s"))
+
+    @check("valid")
+    def uses_condition(self):
+        """project uses at least one condition"""
+        project = json.loads(File("project.json").read())
+        
+        # Search project scripts for an if or if/else block.
+        if not project_contains_keywords(project, ["doIf", "doIfElse"]):
+            raise Error("No conditions found, 1 required.")
+
+    @check("valid")
+    def uses_loop(self):
+        """project uses at least one loop"""
+        project = json.loads(File("project.json").read())
+        
+        # Search project scripts for a repeat, repeat until, or forever block.
+        if not project_contains_keywords(project, ["doRepeat", "doUntil", "doForever"]):
+            raise Error("No loops found, 1 required.")
+
+    @check("valid")
+    def uses_variable(self):
+        """project uses at least one variable"""
+        project = json.loads(File("project.json").read())
+
+        # Look for global variables.
+        if "variables" in project and len(project["variables"]) > 0:
+            return
+        
+        # Look for local-to-sprite variables.
+        for child in project["children"]:
+            if "variables" in child and len(child["variables"]) > 0:
+                return
+        
+        # If we've reached this point, no variable found.
+        raise Error("No variables found, 1 required.")
+
+    @check("valid")
+    def uses_sound(self):
+        """project uses at least one sound"""
+        project = json.loads(File("project.json").read())
+
+        # Search scripts for a sound block.
+        keywords = ["playSound:", "doPlaySoundAndWait", "playDrum", "noteOn:duration:elapsed:from:"]
+        if not project_contains_keywords(project, keywords):
+            raise Error("No sounds found, 1 required.")
+
+def project_contains_keywords(project, keywords):
+    """Returns True if project contains at least one of the keywords."""
+
+    # Iterate over all sprites and backdrops.
+    for child in project["children"]:
+        if "scripts" not in child:
+            continue
+        
+        # Perform a DFS on each script looking for keywords.
+        for script in child["scripts"]:
+            if contains(script, keywords):
+                return True
```

Comment:
Similarly, could probably do `any(contains(script, keywords) for script in child.get("scripts", []))`